hydradx/tests/test_otc.py

```diff
@@ -0,0 +1,337 @@
+        raise
+    if agent.holdings["USDT"] != 1070:
+        raise
+    if agent.holdings["DOT"] != 90:
+        raise
+
+
+def test_buy():
+    # price is USDT/DOT
+    otc = OTC('DOT', 'USDT', 100, 7)
+    agent = Agent(holdings={"USDT": 1000, "DOT": 100})
+    otc.buy(agent, 70)  # should buy 70 USDT for 10 DOT
+    if otc.buy_amount != 10:
+        raise
+    if otc.sell_amount != 30:
+        raise
+    if agent.holdings["USDT"] != 1070:
+        raise
+    if agent.holdings["DOT"] != 90:
+        raise
+
+
+def test_validate():
+    otc = OTC('DOT', 'USDT', 100, 7)
+    if not otc.validate():
+        raise
+    otc = OTC('DOT', 'DOT', 100, 7)
+    if otc.validate():
+        raise
+    otc = OTC('DOT', 'USDT', 100, 0)
+    if otc.validate():
+        raise
+    otc = OTC('DOT', 'USDT', 100, float('inf'))
+    if otc.validate():
+        raise
+    otc = OTC('DOT', 'USDT', 100, -1)
+    if otc.validate():
+        raise
+    otc = OTC('DOT', 'USDT', -1, 7)
+    if otc.validate():
+        raise
+    otc = OTC('DOT', 'USDT', 100, 7)
+    otc.buy_amount = -1
+    if otc.validate():
+        raise
+
+
+def test_sell_fails():
+    otc = OTC('DOT', 'USDT', 100, 7)
+    agent = Agent(holdings={"USDT": 1000, "DOT": 100})
+    with pytest.raises(Exception):
+        otc.sell(agent, 100)  # should fail, too big
+    with pytest.raises(Exception):
+        otc.sell(agent, -1)
+
+
+def test_buy_fails():
+    otc = OTC('DOT', 'USDT', 100, 7)
+    agent = Agent(holdings={"USDT": 1000, "DOT": 100})
+    with pytest.raises(Exception):
+        otc.buy(agent, 110)  # should fail, too big
+    with pytest.raises(Exception):
+        otc.buy(agent, -1)
+
+
+def test_omnipool_settle_otc():
+    tokens = {
+        "HDX": {'liquidity': mpf(1000000), 'LRNA': mpf(100000)},
+        "USDT": {'liquidity': mpf(100000), 'LRNA': mpf(100000)},
+        "DOT": {'liquidity': mpf(10000), 'LRNA': mpf(100000)},
+    }
+
+    init_state = OmnipoolState(
+        tokens=tokens,
+        asset_fee=0.0025,
+        lrna_fee=0.0005,
+        preferred_stablecoin='USDT'
+    )
+
+    # fully execute the OTC order
+    omnipool = init_state.copy()
+    init_sell_amt = 120
+    sell_amt = 120  # USDT we are selling to the Omnipool
+    init_price = 12
+    otc = OTC('DOT', 'USDT', init_sell_amt, init_price)
+    treasury = Agent(holdings={"USDT": 0, "DOT": 0})
+    omnipool_settle_otc(omnipool, otc, treasury, sell_amt)
+    if otc.buy_amount != sell_amt / init_price:
+        raise
+    if otc.sell_amount != init_sell_amt - sell_amt:
+        raise
+    # DOT conserved
+    if (omnipool.liquidity['DOT'] + otc.buy_amount + treasury.holdings['DOT'] != pytest.approx(
+            init_state.liquidity['DOT'], rel=1e-15)):
+        raise
+    # USDT conserved
+    if omnipool.liquidity['USDT'] + treasury.holdings['USDT'] != pytest.approx(init_state.liquidity['USDT'] + init_sell_amt, rel=1e-15):
+        raise
+    if treasury.holdings['DOT'] <= 0:  # treasury should make profit
```

Comment:
I think that correct practice is to put this text (treasury should make profit) into an error message, not a comment.